src/passes/OptimizeInstructions.cpp

```diff
@@ -2815,6 +2818,31 @@ struct OptimizeInstructions
     return false;
   }
 
+  // (x << z) op (y << z)    ==>   (x op y) << z,   op = `|`, `&`, `^`
+  // (x >> z) op (y >> z)    ==>   (x op y) >> z,   op = `|`, `&`, `^`
+  // (x >>> z) op (y >>> z)  ==>   (x op y) >>> z,  op = `|`, `&`, `^`
+  Expression* preserveShift(Binary* curr) {
```

Comment:
The other `preserveOr/And` methods return a bool - they just check if Or/And are preserved. This does more. Perhaps it could be separated into a similar approach, where `preserveShift` returns a bool and the caller does the optimization? If not, having the same name might be confusing.